homeassistant/components/evohome/water_heater.py

```diff
@@ -0,0 +1,107 @@
+"""Support for WaterHeater devices of (EMEA/EU) Honeywell evohome systems."""
+from datetime import datetime, timedelta
+import logging
+
+import requests.exceptions
+
+from homeassistant.components.water_heater import (
+    SUPPORT_OPERATION_MODE,
+    WaterHeaterDevice)
+from homeassistant.const import (
+    STATE_OFF, STATE_ON)
+
+from . import (
+    EvoDevice,
+    CONF_LOCATION_IDX)
+from .climate import (
+    EVO_FOLLOW, EVO_TEMPOVER, EVO_PERMOVER)
+from .const import (
+    DATA_EVOHOME, GWS, TCS)
+
+_LOGGER = logging.getLogger(__name__)
+
+EVO_STATE_TO_HA = {'On': STATE_ON, 'Off': STATE_OFF}
+HA_STATE_TO_EVO = {v: k for k, v in EVO_STATE_TO_HA.items()}
+
+HA_OPMODE_TO_EVO = {STATE_ON: EVO_FOLLOW, STATE_OFF: EVO_PERMOVER}
+
+
+async def async_setup_platform(hass, hass_config, async_add_entities,
+                               discovery_info=None):
+    """Create the DHW controller."""
+    evo_data = hass.data[DATA_EVOHOME]
+
+    client = evo_data['client']
+    loc_idx = evo_data['params'][CONF_LOCATION_IDX]
+
+    # evohomeclient has exposed no means of accessing non-default location
+    # (i.e. loc_idx > 0) other than using a protected member, such as below
+    tcs_obj_ref = client.locations[loc_idx]._gateways[0]._control_systems[0]  # noqa: E501; pylint: disable=protected-access
+
+    _LOGGER.info(
+        "setup(): Found DHW device, id: %s [%s]",
+        tcs_obj_ref.hotwater.zoneId, tcs_obj_ref.hotwater.zone_type)
+
+    dhw = EvoDHW(evo_data, client, tcs_obj_ref.hotwater)
+
+    async_add_entities([dhw], update_before_add=False)
+
+
+class EvoDHW(EvoDevice, WaterHeaterDevice):
+    """Base for a Honeywell evohome DHW controller (aka boiler)."""
+
+    def __init__(self, evo_data, client, obj_ref):
```

Comment:
I'd rename `obj_ref` to something more telling and less general. What is it?

Reply:
It's the reference to the EvoDHW's corresponding object within the client API.

Could we leave it please - it is already `obj_ref`/`self._obj` elsewhere in the integration:
https://github.com/home-assistant/home-assistant/blob/a8e5dbaa7e1e3e3a5fe5bef3bb258bb0851c372e/homeassistant/components/evohome/__init__.py#L147-L151

Reply:
Why not call it `evo_dhw`?

Reply:
I truly think it's less confusing.  It clearly refers to an object and not a data structure, and it wouldn't be to `self`, so therefore the corresponding object in the Client.

Of course, the `super()` would be `evo_object` (or similar) as it is used for Climate entities, and well as WaterHeaters.

Experience of refactoring my own code tells me to leave it as is.  I will happily change `obj_ref` to `evo_dhw`, but can I leave `self._obj` as is?

Reply:
Object is probably the most generic name you can choose in Python, as everything is objects. We should try to convey the use of a variable or its origin or something that makes it stand out, with its name.